text/0000-read-buf.md

```diff
@@ -0,0 +1,631 @@
+        let len = unfilled.len();
+        buf.add_filled(len);
+
+        Ok(())
+    }
+}
+```
+
+An unsafe `Read` implementation:
+
+```rust
+impl Read for TcpStream {
+    fn read_buf(&mut self, buf: &mut ReadBuf<'_>) -> io::Result<()> {
+        unsafe {
+            // Get access to the filled part of the buffer, without initializing it. This method is unsafe; we are
+            // responsible for ensuing that we don't "de-initialize" portions of it that have previously been
+            // initialized.
+            let unfilled: &mut [MaybeUninit<u8>] = buf.unfilled_mut();
+
+            // We're just delegating to the libc read function, which returns an `isize`. The return value indicates
+            // an error if negative and the number of bytes read otherwise.
+            let nread = libc::read(self.fd, unfilled.as_mut_ptr().cast::<libc::c_void>(), unfilled.len());
+
+            if nread < 0 {
+                return Err(io::Error::last_os_error());
+            }
+
+            let nread = nread as usize;
+            // If the read succeeded, tell the buffer that the read-to portion has been initialized. This method is
+            // unsafe; we are responsible for ensuring that this portion of the buffer has actually been initialized.
+            buf.assume_init(nread);
+            // And indicate that we've written the bytes as well. Unlike `assume_initialized`, this method is safe,
+            // and asserts that the written portion of the buffer does not advance beyond the initialized portion of
+            // the buffer. If we didn't call `assume_initialized` above, this call could panic.
+            buf.add_filled(nread);
+
+            Ok(())
+        }
+    }
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+```rust
+/// A wrapper around a byte buffer that is incrementally filled and initialized.
+///
+/// This type is a sort of "double cursor". It tracks three regions in the buffer: a region at the beginning of the
+/// buffer that has been logically filled with data, a region that has been initialized at some point but not yet
+/// logically filled, and a region at the end that is fully uninitialized. The filled region is guaranteed to be a
+/// subset of the initialized region.
+pub struct ReadBuf<'a> {
+    buf: &'a mut [MaybeUninit<u8>],
+    filled: usize,
+    initialized: usize,
+}
+
+impl<'a> ReadBuf<'a> {
+    /// Creates a new `ReadBuf` from a fully initialized buffer.
+    #[inline]
+    pub fn new(buf: &'a mut [u8]) -> ReadBuf<'a> { ... }
+
+    /// Creates a new `ReadBuf` from a fully uninitialized buffer.
+    ///
+    /// Use `assume_initialized` if part of the buffer is known to be already inintialized.
+    #[inline]
+    pub fn uninit(buf: &'a mut [MaybeUninit<u8>]) -> ReadBuf<'a> { ... }
+
+    /// Returns the full size of the buffer.
+    #[inline]
+    pub fn len(&self) -> usize { ... }
```

Comment:
Picked up by clippy and literally first blocker I had trying out this API, there should be `pub fn is_empty(&self) -> bool` to go along with this.

Reply:
Though, actually it might be a good thing that this doesn't exist since the actual check that I need is `buf.remaining() == 0`. Maybe this is a sign that `len` is too generic a name when there are so many different lengths to use?

Reply:
`capacity` seems better

Reply:
Updated